nums/core/array/application.py

```diff
@@ -1207,3 +1206,79 @@ def column_stack(self, tup, axis_block_size=None):
             else:
                 arrays.append(self.atleast_2d(arr))
         return self.concatenate(arrays, 1, axis_block_size=axis_block_size)
+
+    def sort(self, arr):
+        pivot_oids = arr.flattened_oids()
+        num_blocks = len(pivot_oids)
+
+        # If the BlockArray to be sorted is small enough, default to serial sort
+        if num_blocks == 1:
+            first_oid = self.cm.get(pivot_oids[0])
+            oid = self.cm.sort(
+                first_oid,
+                syskwargs={
+                    "grid_entry": (0,),
+                    "grid_shape": (1,),
+                    "options": {"num_returns": 1},
+                },
+            )
+            return BlockArray.from_oid(oid, arr.block_shape, arr.dtype, self.cm)
+
+        pivot_oids.pop()
+
+        # Sample the pivots to sort buckets around
+        pivots = []
+        for i, arr_oid in enumerate(pivot_oids):
+            oid = self.cm.sample_pivots(
+                arr_oid,
+                syskwargs={
+                    "grid_entry": (i,),
+                    "grid_shape": (num_blocks - 1,),
+                    "options": {"num_returns": 1},
+                },
+            )
+            pivots.append(BlockArray.from_oid(oid, (1,), arr.dtype, self.cm))
+
+        new_pivots = self.concatenate(pivots, 0, axis_block_size=num_blocks - 1)
+
+        # Assume that the number of pivots is small enough to fit within a single BlockArray
+        sorted_pivots_oid = self.cm.sort(
+            new_pivots.blocks[0].oid,
+            syskwargs={
+                "grid_entry": (0,),
+                "grid_shape": (1,),
+                "options": {"num_returns": 1},
+            },
+        )
+
+        mapped_sorted = np.empty([num_blocks, num_blocks], dtype=object)
+        reduce_sorted = []
+
+        arr_oids = arr.flattened_oids()
+        for i in range(num_blocks):
+            mapped_sorted[i] = self.cm.map_sort(
+                arr_oids[i],
+                sorted_pivots_oid,
+                syskwargs={
+                    "grid_entry": (0,),
+                    "grid_shape": (1,),
+                    "options": {"num_returns": num_blocks},
+                },
+            )
+
```

Comment:
The block decomposition of the output here should be the same as the input. You can initialize an empty output BlockArray and assign the output oids to each block of the output block array. Might be worth double checking that the sort op is similar semantically to shuffle (for this to hold).